### PR TITLE
Allow res pages in embedded webview

### LINKED ISSUE #3083: [Bug] Allow res:// error pages to be displayed in embedded webview
msal 4.39

ICM follow up 
res:// scheme is an IE internal scheme for displaying errors. It should be allowed to improve supporatibility.

### PR DESCRIPTION
**Pull request by bgavrilMS**
Fixes #3083

**Changes proposed in this request**

**Testing**

**Performance impact**

### CODE REVIEW

File: src/client/Microsoft.Identity.Client/Platforms/Features/WebView2WebUi/WinFormsPanelWithWebView2.cs [hunk 1]
```diff
@@ -214,10 +214,12 @@ private bool CheckForEndUrl(Uri url)
             if (!readyToClose &&
                 !url.Scheme.Equals("https", StringComparison.OrdinalIgnoreCase) &&
```
Comment (SameerK-MSFT) [inline src/client/Microsoft.Identity.Client/Platforms/Features/WebView2WebUi/WinFormsPanelWithWebView2.cs:215]: minor: can we have url extension where this validation is performed. So the same code in two files will be centralized and any further changes will be at one place.

**Review by SameerK-MSFT — APPROVED** [review]

File: src/client/Microsoft.Identity.Client/Platforms/Features/WebView2WebUi/WinFormsPanelWithWebView2.cs [hunk 2]
```diff
@@ -214,10 +214,12 @@ private bool CheckForEndUrl(Uri url)
...
                 !url.AbsoluteUri.Equals("about:blank", StringComparison.OrdinalIgnoreCase) &&
-                !url.Scheme.Equals("javascript", StringComparison.OrdinalIgnoreCase))
+                !url.Scheme.Equals("javascript", StringComparison.OrdinalIgnoreCase) &&
+                !url.Scheme.Equals("res", StringComparison.OrdinalIgnoreCase) // IE error pages
+                )
             {
 
-                _logger.Error($"Redirection to non-HTTPS scheme ({url.Scheme}) found! WebView will fail...");
+                _logger.Error($"Redirection to non-HTTPS scheme ({url}) found! WebView2 will fail...");
```
Comment (pmaytak) [inline src/client/Microsoft.Identity.Client/Platforms/Features/WebView2WebUi/WinFormsPanelWithWebView2.cs:222]: nit: maybe add "[WebView2]" log prefix here (or if there's already one that we use.
Also maybe reword a bit - change 'scheme' to 'url`, since that's what's being printed? URL is non-PII too, right?

File: src/client/Microsoft.Identity.Client/Platforms/Features/WinFormsLegacyWebUi/WindowsFormsWebAuthenticationDialogBase.cs [hunk 3]
```diff
@@ -210,10 +210,13 @@ private bool CheckForClosingUrl(Uri url, byte[] postData = null)
             }
 
             if (!readyToClose && !url.Scheme.Equals("https", StringComparison.OrdinalIgnoreCase) &&
-                !url.AbsoluteUri.Equals("about:blank", StringComparison.OrdinalIgnoreCase) && !url.Scheme.Equals("javascript", StringComparison.OrdinalIgnoreCase))
+                !url.AbsoluteUri.Equals("about:blank", StringComparison.OrdinalIgnoreCase) &&
+                !url.Scheme.Equals("javascript", StringComparison.OrdinalIgnoreCase) &&
+                url.Scheme.Equals("res", StringComparison.OrdinalIgnoreCase) // IE error pages
```
Comment (pmaytak) [inline src/client/Microsoft.Identity.Client/Platforms/Features/WinFormsLegacyWebUi/WindowsFormsWebAuthenticationDialogBase.cs:215]: Missing `!`?
```suggestion
                !url.Scheme.Equals("res", StringComparison.OrdinalIgnoreCase) // IE error pages
```

File: src/client/Microsoft.Identity.Client/Platforms/Features/WinFormsLegacyWebUi/WindowsFormsWebAuthenticationDialogBase.cs [hunk 4]
```diff
@@ -210,10 +210,13 @@ private bool CheckForClosingUrl(Uri url, byte[] postData = null)
...
+                )
             {
                 RequestContext.Logger.Error(string.Format(CultureInfo.InvariantCulture,
-                    "[Legacy WebView] Redirection to non-HTTPS scheme ({0}) found! WebView will fail...", url.Scheme));
+                    "[Legacy WebView] Redirection to non-HTTPS scheme ({0}) found! WebView1 will fail...", url));
```
Comment (pmaytak) [inline src/client/Microsoft.Identity.Client/Platforms/Features/WinFormsLegacyWebUi/WindowsFormsWebAuthenticationDialogBase.cs:219]: Maybe reword a bit - same as comment in the other file.

**Review by pmaytak — CHANGES_REQUESTED** [review]

**Comment by bgavrilMS** [comment]
Thanks for the careful review @pmaytak @SameerK-MSFT - added a few fixes.

File: src/client/Microsoft.Identity.Client/Platforms/Features/WebView2WebUi/WinFormsPanelWithWebView2.cs [hunk 5]
```diff
@@ -214,10 +214,12 @@ private bool CheckForEndUrl(Uri url)
             if (!readyToClose &&
                 !url.Scheme.Equals("https", StringComparison.OrdinalIgnoreCase) &&
                 !url.AbsoluteUri.Equals("about:blank", StringComparison.OrdinalIgnoreCase) &&
-                !url.Scheme.Equals("javascript", StringComparison.OrdinalIgnoreCase))
+                !url.Scheme.Equals("javascript", StringComparison.OrdinalIgnoreCase) &&
+                !url.Scheme.Equals("res", StringComparison.OrdinalIgnoreCase) // IE error pages
```
Comment (pmaytak) [inline src/client/Microsoft.Identity.Client/Platforms/Features/WebView2WebUi/WinFormsPanelWithWebView2.cs:218]: Also instead call `!EmbeddedUiCommon.IsAllowedIeOrEdgeAuthorizationUri(url)` here?

File: src/client/Microsoft.Identity.Client/Platforms/Features/WinFormsLegacyWebUi/WindowsFormsWebAuthenticationDialogBase.cs [hunk 6]
```diff
@@ -209,11 +209,10 @@ private bool CheckForClosingUrl(Uri url, byte[] postData = null)
                 readyToClose = true;
             }
 
-            if (!readyToClose && !url.Scheme.Equals("https", StringComparison.OrdinalIgnoreCase) &&
-                !url.AbsoluteUri.Equals("about:blank", StringComparison.OrdinalIgnoreCase) && !url.Scheme.Equals("javascript", StringComparison.OrdinalIgnoreCase))
+            if (!readyToClose && EmbeddedUiCommon.IsAllowedIeOrEdgeAuthorizationUri(url)) // IE error pages
```
Comment (pmaytak) [inline src/client/Microsoft.Identity.Client/Platforms/Features/WinFormsLegacyWebUi/WindowsFormsWebAuthenticationDialogBase.cs:212]: I think this should be `!`EmbeddedUiCommon.IsAllowedIeOrEdgeAuthorizationUri(url) because the logic was changed from NOT ANDs to ORs?

Comment (bgavrilMS) [reply src/client/Microsoft.Identity.Client/Platforms/Features/WinFormsLegacyWebUi/WindowsFormsWebAuthenticationDialogBase.cs]: Yes, should have tested that.

**Review by pmaytak — APPROVED** [review]
Good comment changes.
